### PR TITLE
[SEARCH-1263]Date caching locale and timezone format fix to comply wi…

### PR DESCRIPTION
**Pull request by alessandrobenedetti**
This is the key method which is used from the Alfresco Search Services module to retrieve a Solr date formatter to then build date range queries.
It must be fixed to comply with the ISO 8601 standard.

### CODE REVIEW

File: src/test/java/org/alfresco/util/CachingDateFormatTest.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package org.alfresco.util;
```
Comment (tpage-alfresco) [inline src/test/java/org/alfresco/util/CachingDateFormatTest.java:19]: I think we usually have copyright headers - even for test files.

File: src/test/java/org/alfresco/util/CachingDateFormatTest.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        
+        String formattedDate = solrDatetimeFormat.format(shanghaiDate);
+        
+        assertThat(formattedDate,is("2018-04-01T02:00:00.000Z"));
+    }
+
+    @Test
+    public void solrDatetimeFormat_LocaleNotWestern_shouldReturnISO8601DateString()
+    {
+        Locale.setDefault(new Locale("hi", "IN"));
```
Comment (tpage-alfresco) [inline src/test/java/org/alfresco/util/CachingDateFormatTest.java]: Will this affect all subsequent tests?  Perhaps we should set it back after the test?  (Also, out of interest, is this Hindi? Maybe it's worth saying why you're picking this locale - e.g. it defaults to a different number system)

**Review by tpage-alfresco — COMMENTED** [review]
A couple of minor points about the test file, but otherwise looks good to me :+1:

**Comment by tpage-alfresco** [comment]
Also as a general point about the original issue: Would it be reasonable to test all locales, rather than picking a random locale for the tests. It seems weird to have tests that we *expect* to fail intermittently if there's a problem.  If the tests would take too long to run against all locales then maybe we could at least make it obvious in the logs that a locale (and timezone) had been picked randomly for the test run (and what was picked)?

**Comment by alessandrobenedetti** [comment]
I will pick up the suggested changes and update the pull request.
In response to the last point, this is how actually Apache Solr tests work (and Alfresco extends).
I disagree with random tests and I agree with you, there shouldn't be a random parameter that make tests intermittent but all scenarios (such different locales should be addressed separately).

Given that, having random seed for exploring different scenarios is still better than nothing (and allowed us to identify this bug).

On the last point the tests will already tell you those details :

NOTE: reproduce with: ant test  -Dtestcase=DistributedTimeSeriesStreamTest -Dtests.method=testTimeSeries -Dtests.seed=4A9FBEDD558DFD49 **-Dtests.locale=de-DE** **-Dtests.timezone=America/Thule** -Dtests.asserts=true -Dtests.file.encoding=UTF-8

**Comment by tpage-alfresco** [comment]
Great - thanks Alessandro. I didn't realise the randomisation wasn't in our codebase.  I wouldn't worry about changing it along with this change then, maybe something to discuss with the team during a technical AOB (e.g. it might be useful to ask 'Are there some locales and timezones that we expect to cause problems, and are there some that we can ignore?' or 'How much extra build time would it cost to support an extra locale/timezone?')

File: src/test/java/org/alfresco/util/CachingDateFormatTest.java (hunk 1 above)
Comment (alessandrobenedetti) [reply src/test/java/org/alfresco/util/CachingDateFormatTest.java:19]: Done!

File: src/test/java/org/alfresco/util/CachingDateFormatTest.java (hunk 2 above)
Comment (alessandrobenedetti) [reply src/test/java/org/alfresco/util/CachingDateFormatTest.java]: Done, you are absolutely right!

**Comment by alessandrobenedetti** [comment]
I have pushed the changes discussed!
In addition a big +1 on having a careful discussion on where Locales can break things!

**Review by tpage-alfresco — APPROVED** [review]
LGTM

**Pull request merged**